src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -966,6 +969,8 @@ public static PropertyFetch FetcherForProperty(Type? type, string propertyName)
 
                     private sealed class RefTypedFetchProperty<TObject, TProperty> : PropertyFetch
                     {
+                        [DynamicDependency("get_TraceStateString", typeof(Activity))]
```

Comment:
I'm not understanding the reasoning behind these DynamicDependencies. Can you explain?

Reply:
The `FetcherForProperty` method reflects over properties on `System.Diagnostics.Activity`, depending on what properties are passed to an `EventListener`, e.g. see [this usage](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/tests/TrimmingTests/EventSourcePropertyFetchTest.cs#L23-L47). I believe the getters for these properties are used to report activity on whatever the listener is listening for. 

The [example](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/tests/TrimmingTests/EventSourcePropertyFetchTest.cs#L23-L47) passes in all non-static* properties on `typeof(Activity)` to a listener, and it turns out that the getters for `TraceStateString` and `IsAllDataRequested` are the only two that are trimmed out. This would cause a [NullReferenceException ](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L977)in a trimmed app with this usage pattern.

---
*A problem with static properties is discussed here - https://github.com/dotnet/runtime/pull/38216#discussion_r443764227.

Reply:
This looks like an annotation that makes the trimming test pass, but doesn't actually make DiagnosticSource trimming safe.

There seems to be code in DiagnosticSource that reflects on properties passed as a text string into the source. The text string is `activityProps` in the `EventSourcePropertyFetchTest.cs`.

If we were to add a property on a class there that is not otherwise used, the reflection is going to crash and fail.

I think if we were to replace the 

```csharp
var obj = new { DummyProp = "val" };
```

in the test with

```csharp
var obj = new NeverUsedProperties();
// class NeverUsedProperties { public string DummyProp { get; set; } }
```

This is actually going to crash because linker will trim `DummyProp` away, but the `activityProps` string is still going to refer to it.

We might want to just annotate DiagnosticSourceEventSource as not trimming safe and revisit in .NET 6.

Reply:
Well, maybe it won't crash, just silently not work, because the activityProps string is already referencing a bunch of non-existent properties that DiagnosticSource is complaining about:

```
Property Links not found on System.Diagnostics.Activity
Property IsAllDataRequested not found on System.Diagnostics.Activity
Property Kind not found on System.Diagnostics.Activity
Property DisplayName not found on System.Diagnostics.Activity
Property Source not found on System.Diagnostics.Activity
Property Events not found on System.Diagnostics.Activity
Property Context not found on System.Diagnostics.Activity
```

Reply:
FWIW - In an earlier iteration of this PR, I placed `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)]` on the `type` parameter in [this method](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L895), which I thought would preserve the needed properties and their accessors. That didn't work. Is the expected behavior with `DynamicallyAccessedMemberTypes.PublicProperties` that the accessors will be preserved?

Reply:
> Is the expected behavior with DynamicallyAccessedMemberTypes.PublicProperties that the accessors will be preserved?

You need to add annotations all the way up until you get to a concrete type. The code that calls into `PropertyFetch.FetcherForProperty` doesn't call it with a concrete type:

https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L858-L864

Here the linker has no way of understanding what `obj?.GetType()` will return since `obj` is of type `object`.

I agree it makes sense to mark `DiagnosticSourceEventSource` as not trimming safe for .NET 5 and log an issue to address it in the future.

Reply:
Annotating `type` would do the right thing for the code in that method, but I assume you got a warning from [this line](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L862) because linker wouldn't be able to figure out where that `objType` comes from.

I tried to manually trace that back, but I gave up after I saw the source of the `objType` ([this line](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L859)) requires tracing through a KeyValuePair, delegate, interface call, and I still couldn't figure out what types could be stored in the `object`-typed parameter. The only safe thing to do for linker in that spot would be to preserve all property on everything deriving from `System.Object` and that's kind of counter productive to trimming.

Reply:
I see.

I'll pull out related changes from this PR. I've logged an issue to mark the method as unsafe for trimming - https://github.com/dotnet/runtime/issues/38404. Will address in a follow up.